### PR TITLE
chore(telemetry): track tracer configurations

### PR DESCRIPTION
**Pull request by mabdinur**
Uses instrumentation telemetry to track the value of the following configurations on application start up:
  ```
  DD_INSTRUMENTATION_TELEMETRY_ENABLED
  DD_LOGS_INJECTION
  DD_TRACE_128_BIT_TRACEID_GENERATION_ENABLED
  DD_TRACE_128_BIT_TRACEID_LOGGING_ENABLED
  DD_TRACE_ANALYTICS_ENABLED
  DD_TRACE_CLIENT_IP_ENABLED
  DD_TRACE_COMPUTE_STATS
  DD_TRACE_DEBUG
  DD_TRACE_ENABLED
  DD_TRACE_HEALTH_METRICS_ENABLED
  DD_TRACE_OBFUSCATION_QUERY_STRING_PATTERN
  ```

Notes: 
 - Configurations that are set after an application starts up will not be tracked.  Support for configurations that changed in code will be introduced in a future PR.
 - Some tracer components were refactored to avoid circular imports. If these changes are too difficult to follow then I can open separate PRs with those changes.
 - Support for tracking sampling rules and remote config will be added in a future PR.

## Checklist

- [x] Change(s) are motivated and described in the PR description.
- [x] Testing strategy is described if automated tests are not included in the PR.
- [x] Risk is outlined (performance impact, potential for breakage, maintainability, etc).
- [x] Change is maintainable (easy to change, telemetry, documentation).
- [x] [Library release note guidelines](https://ddtrace.readthedocs.io/en/stable/releasenotes.html) are followed. If no release note is required, add label `changelog/no-changelog`.
- [x] Documentation is included (in-code, generated user docs, [public corp docs](https://github.com/DataDog/documentation/)).
- [x] Backport labels are set (if [applicable](https://ddtrace.readthedocs.io/en/latest/contributing.html#backporting))

## Reviewer Checklist

- [x] Title is accurate.
- [x] No unnecessary changes are introduced.
- [x] Description motivates each change.
- [x] Avoids breaking [API](https://ddtrace.readthedocs.io/en/stable/versioning.html#interfaces) changes unless absolutely necessary.
- [x] Testing strategy adequately addresses listed risk(s).
- [x] Change is maintainable (easy to change, telemetry, documentation).
- [x] Release note makes sense to a user of the library.
- [x] Reviewer has explicitly acknowledged and discussed the performance implications of this PR as reported in the benchmarks PR comment.
- [x] Backport labels are set in a manner that is consistent with the [release branch maintenance policy](https://ddtrace.readthedocs.io/en/latest/contributing.html#backporting)

### CODE REVIEW

File: ddtrace/internal/_encoding.pyi [hunk 1]
```diff
@@ -5,6 +5,7 @@ from typing import Union
 
 from ddtrace.span import Span
 
+
```
Comment (mabdinur) [inline ddtrace/internal/_encoding.pyi:8]:
```suggestion
```

File: ddtrace/internal/constants.py [hunk 2]
```diff
@@ -15,6 +15,7 @@
 W3C_TRACEPARENT_KEY = "traceparent"
 W3C_TRACESTATE_ORIGIN_KEY = "o"
 W3C_TRACESTATE_SAMPLING_PRIORITY_KEY = "s"
+SAMPLING_DECISION_TRACE_TAG_KEY = "_dd.p.dm"
```
Comment (mabdinur) [inline ddtrace/internal/constants.py:18]: This change is required to avoid circular imports

**Review by majorgreys — APPROVED** [review]

File: ddtrace/settings/config.py [hunk 3]
```diff
@@ -193,6 +193,9 @@ def __init__(self):
         # use a dict as underlying storing mechanism
         self._config = {}
 
+        self.debug_mode = asbool(os.getenv("DD_TRACE_DEBUG", default=False))
+        self.call_basic_config = asbool(os.environ.get("DD_CALL_BASIC_CONFIG", "false"))
```
Comment (mabdinur) [inline ddtrace/settings/config.py:197]:
```suggestion
        self._debug_mode = asbool(os.getenv("DD_TRACE_DEBUG", default=False))
        self._call_basic_config = asbool(os.environ.get("DD_CALL_BASIC_CONFIG", "false"))
```

File: ddtrace/tracer.py [hunk 4]
```diff
@@ -79,14 +79,12 @@
...
 
-debug_mode = asbool(os.getenv("DD_TRACE_DEBUG", default=False))
-call_basic_config = asbool(os.environ.get("DD_CALL_BASIC_CONFIG", "false"))
 
 DD_LOG_FORMAT = "%(asctime)s %(levelname)s [%(name)s] [%(filename)s:%(lineno)d] {}- %(message)s".format(
     "[dd.service=%(dd.service)s dd.env=%(dd.env)s dd.version=%(dd.version)s"
     " dd.trace_id=%(dd.trace_id)s dd.span_id=%(dd.span_id)s] "
 )
-if debug_mode and not hasHandlers(log) and call_basic_config:
+if config.debug_mode and not hasHandlers(log) and config.call_basic_config:
```
Comment (mabdinur) [inline ddtrace/tracer.py:87]:
```suggestion
if config._debug_mode and not hasHandlers(log) and config._call_basic_config:
```

File: ddtrace/tracer.py [hunk 5]
```diff
@@ -542,7 +540,7 @@ def configure(
         self._generate_diagnostic_logs()
 
     def _generate_diagnostic_logs(self):
-        if debug_mode or asbool(environ.get("DD_TRACE_STARTUP_LOGS", False)):
+        if config.debug_mode or asbool(environ.get("DD_TRACE_STARTUP_LOGS", False)):
```
Comment (mabdinur) [inline ddtrace/tracer.py:543]:
```suggestion
        if config._debug_mode or asbool(environ.get("DD_TRACE_STARTUP_LOGS", False)):
```

File: ddtrace/bootstrap/sitecustomize.py [hunk 6]
```diff
@@ -36,8 +35,7 @@
 # upon initializing it the first time.
 # See https://github.com/python/cpython/blob/112e4afd582515fcdcc0cde5012a4866e5cfda12/Lib/logging/__init__.py#L1550
 # Debug mode from the tracer will do a basicConfig so only need to do this otherwise
-call_basic_config = asbool(os.environ.get("DD_CALL_BASIC_CONFIG", "false"))
-if not debug_mode and call_basic_config:
+if not config.debug_mode and config.call_basic_config:
```
Comment (mabdinur) [inline ddtrace/bootstrap/sitecustomize.py:38]:
```suggestion
if not config._debug_mode and config._call_basic_config:
```

File: ddtrace/commands/ddtrace_run.py [hunk 7]
```diff
@@ -92,9 +91,7 @@ def main():
     if args.profiling:
         os.environ["DD_PROFILING_ENABLED"] = "true"
 
-    debug_mode = args.debug or asbool(os.getenv("DD_TRACE_DEBUG", default=False))
-
-    if debug_mode:
+    if args.debug or ddtrace.config.debug_mode:
```
Comment (mabdinur) [inline ddtrace/commands/ddtrace_run.py:94]:
```suggestion
    if args.debug or ddtrace.config._debug_mode:
```

File: ddtrace/internal/telemetry/writer.py [hunk 8]
```diff
@@ -276,7 +287,24 @@ def _app_started_event(self):
                 (TELEMETRY_PROPAGATION_STYLE_EXTRACT, ",".join(config._propagation_style_extract), "unknown"),
                 ("ddtrace_bootstrapped", config._ddtrace_bootstrapped, "unknown"),
                 ("ddtrace_auto_used", "ddtrace.auto" in sys.modules, "unknown"),
-                ("otel_enabled", config._otel_enabled, "unknown"),
+                (TELEMETRY_RUNTIMEMETRICS_ENABLED, config._runtime_metrics_enabled, "unknown"),
+                (TELEMETRY_TRACE_DEBUG, config.debug_mode, "unknown"),
```
Comment (mabdinur) [inline ddtrace/internal/telemetry/writer.py:291]:
```suggestion
                (TELEMETRY_TRACE_DEBUG, config._debug_mode, "unknown"),
```

File: ddtrace/internal/telemetry/constants.py [hunk 9]
```diff
@@ -8,15 +8,29 @@
...
 TELEMETRY_ASM_ENABLED = "DD_APPSEC_ENABLED"
 TELEMETRY_DYNAMIC_INSTRUMENTATION_ENABLED = "DD_DYNAMIC_INSTRUMENTATION_ENABLED"
 TELEMETRY_EXCEPTION_DEBUGGING_ENABLED = "DD_EXCEPTION_DEBUGGING_ENABLED"
 TELEMETRY_DSM_ENABLED = "DD_DATA_STREAMS_ENABLED"
+
+
+# Tracing Features
+
+TELEMETRY_TRACE_DEBUG = "DD_TRACE_DEBUG"
+TELEMETRY_ANALYTICS_ENABLED = "DD_TRACE_ANALYTICS_ENABLED"
```
Comment (mabdinur) [inline ddtrace/internal/telemetry/constants.py:22]:
```suggestion
TELEMETRY_ANALYTICS_ENABLED = "DD_TRACE_ANALYTICS_ENABLED"
TELEMETRY_CALL_BASIC_CONFIG = "DD_CALL_BASIC_CONFIG"
```

File: ddtrace/internal/telemetry/writer.py [hunk 10]
```diff
@@ -30,14 +30,25 @@
 from ..utils.formats import asbool
 from ..utils.time import StopWatch
 from ..utils.version import _pep440_to_semver
+from .constants import TELEMETRY_128_BIT_TRACEID_GENERATION_ENABLED
+from .constants import TELEMETRY_128_BIT_TRACEID_LOGGING_ENABLED
+from .constants import TELEMETRY_ANALYTICS_ENABLED
 from .constants import TELEMETRY_ASM_ENABLED
+from .constants import TELEMETRY_CLIENT_IP_ENABLED
```
Comment (mabdinur) [inline ddtrace/internal/telemetry/writer.py:37]:
```suggestion
from .constants import TELEMETRY_CALL_BASIC_CONFIG
from .constants import TELEMETRY_CLIENT_IP_ENABLED
```

File: ddtrace/internal/telemetry/writer.py [hunk 11]
```diff
@@ -276,7 +287,24 @@ def _app_started_event(self):
                 (TELEMETRY_PROPAGATION_STYLE_EXTRACT, ",".join(config._propagation_style_extract), "unknown"),
                 ("ddtrace_bootstrapped", config._ddtrace_bootstrapped, "unknown"),
                 ("ddtrace_auto_used", "ddtrace.auto" in sys.modules, "unknown"),
-                ("otel_enabled", config._otel_enabled, "unknown"),
+                (TELEMETRY_RUNTIMEMETRICS_ENABLED, config._runtime_metrics_enabled, "unknown"),
+                (TELEMETRY_TRACE_DEBUG, config.debug_mode, "unknown"),
+                (TELEMETRY_ENABLED, config._telemetry_enabled, "unknown"),
+                (TELEMETRY_ANALYTICS_ENABLED, config.analytics_enabled, "unknown"),
+                (TELEMETRY_CLIENT_IP_ENABLED, config.client_ip_header, "unknown"),
+                (TELEMETRY_LOGS_INJECTION_ENABLED, config.logs_injection, "unknown"),
```
Comment (mabdinur) [inline ddtrace/internal/telemetry/writer.py:295]:
```suggestion
                (TELEMETRY_LOGS_INJECTION_ENABLED, config.logs_injection, "unknown"),
                (TELEMETRY_CALL_BASIC_CONFIG, config._call_basic_config, "unknown"),
```

File: tests/telemetry/test_writer.py [hunk 12]
```diff
@@ -76,14 +74,35 @@ def test_app_started_event(telemetry_writer, test_agent_session, mock_time):
             {"name": "DD_DATA_STREAMS_ENABLED", "origin": "unknown", "value": False},
```
Comment (mabdinur) [inline tests/telemetry/test_writer.py:74]:
```suggestion
            {"name": "DD_CALL_BASIC_CONFIG", "origin": "unknown", "value": False},
            {"name": "DD_DATA_STREAMS_ENABLED", "origin": "unknown", "value": False},
```

File: tests/telemetry/test_writer.py [hunk 13]
```diff
@@ -116,16 +135,26 @@ def test_app_started_event_configuration_override(test_agent_session, run_python
...
-    env["DD_PROFILING_ENABLED"] = "true"
-    env["DD_DYNAMIC_INSTRUMENTATION_ENABLED"] = "true"
-    env[TELEMETRY_PROPAGATION_STYLE_EXTRACT] = "b3multi"
-    env[TELEMETRY_PROPAGATION_STYLE_INJECT] = "datadog"
-    env["DD_TRACE_OTEL_ENABLED"] = "true"
-    env["DD_RUNTIME_METRICS_ENABLED"] = "true"
-    env["DD_EXCEPTION_DEBUGGING_ENABLED"] = "true"
+    env["DD_EXCEPTION_DEBUGGING_ENABLED"] = "True"
+    env["DD_INSTRUMENTATION_TELEMETRY_ENABLED"] = "True"
+    env["DD_LOGS_INJECTION"] = "True"
```
Comment (mabdinur) [inline tests/telemetry/test_writer.py:140]:
```suggestion
    env["DD_LOGS_INJECTION"] = "True"
    env["DD_CALL_BASIC_CONFIG"] = "True"
```

File: tests/telemetry/test_writer.py [hunk 14]
```diff
@@ -138,19 +167,27 @@ def test_app_started_event_configuration_override(test_agent_session, run_python
     events[0]["payload"]["configuration"].sort(key=lambda c: c["name"])
 
     assert events[0]["payload"]["configuration"] == [
-        {"name": "DD_APPSEC_ENABLED", "origin": "unknown", "value": True},
-        {"name": "DD_DATA_STREAMS_ENABLED", "origin": "unknown", "value": True},
-        {"name": "DD_DYNAMIC_INSTRUMENTATION_ENABLED", "origin": "unknown", "value": True},
+        {"name": "DD_APPSEC_ENABLED", "origin": "unknown", "value": False},
+        {"name": "DD_DATA_STREAMS_ENABLED", "origin": "unknown", "value": False},
```
Comment (mabdinur) [inline tests/telemetry/test_writer.py:171]:
```suggestion
        {"name": "DD_CALL_BASIC_CONFIG", "origin": "unknown", "value": True},
        {"name": "DD_DATA_STREAMS_ENABLED", "origin": "unknown", "value": False},
```

File: tests/telemetry/test_writer.py [hunk 15]
```diff
@@ -73,17 +71,39 @@ def test_app_started_event(telemetry_writer, test_agent_session, mock_time):
...
+                '(?:ed|ature)?|auth(?:entication|orization)?)(?:(?:\\s|%20)*(?:=|%3D)[^&]+|(?:"|'
+                '%22)(?:\\s|%20)*(?::|%3A)(?:\\s|%20)*(?:"|%22)(?:%2[^2]|%[^2]|[^"%])+(?:"|%22))|'
+                "bearer(?:\\s|%20)+[a-z0-9\\._\\-]|token(?::|%3A)[a-z0-9]{13}|gh[opsu]_[0-9a-zA-Z]{36}|"
+                "ey[I-L](?:[\\w=-]|%3D)+\\.ey[I-L](?:[\\w=-]|%3D)+(?:\\.(?:[\\w.+\\/=-]|%3D|%2F|%2B)+)?|["
+                "\\-]{5}BEGIN(?:[a-z\\s]|%20)+PRIVATE(?:\\s|%20)KEY[\\-]{5}[^\\-]+[\\-]{5}END(?:[a-z\\s]|%"
+                "20)+PRIVATE(?:\\s|%20)KEY|ssh-rsa(?:\\s|%20)*(?:[a-z0-9\\/\\.+]|%2F|%5C|%2B){100,}",
+            },
+            {"name": "DD_TRACE_OTEL_ENABLED", "origin": "unknown", "value": False},
+            {"name": "DD_TRACE_PROPAGATION_STYLE_EXTRACT", "origin": "unknown", "value": "datadog"},
+            {"name": "DD_TRACE_PROPAGATION_STYLE_INJECT", "origin": "unknown", "value": "datadog"},
```
Comment (mabdinur) [inline tests/telemetry/test_writer.py:104]:
```suggestion
            {"name": "DD_TRACE_PROPAGATION_STYLE_EXTRACT", "origin": "unknown", "value": "tracecontext ,datadog"},
            {"name": "DD_TRACE_PROPAGATION_STYLE_INJECT", "origin": "unknown", "value": "tracecontext ,datadog"},
```

File: tests/telemetry/test_writer.py [hunk 16]
```diff
@@ -73,17 +71,39 @@ def test_app_started_event(telemetry_writer, test_agent_session, mock_time):
...
+                '(?:ed|ature)?|auth(?:entication|orization)?)(?:(?:\\s|%20)*(?:=|%3D)[^&]+|(?:"|'
+                '%22)(?:\\s|%20)*(?::|%3A)(?:\\s|%20)*(?:"|%22)(?:%2[^2]|%[^2]|[^"%])+(?:"|%22))|'
+                "bearer(?:\\s|%20)+[a-z0-9\\._\\-]|token(?::|%3A)[a-z0-9]{13}|gh[opsu]_[0-9a-zA-Z]{36}|"
+                "ey[I-L](?:[\\w=-]|%3D)+\\.ey[I-L](?:[\\w=-]|%3D)+(?:\\.(?:[\\w.+\\/=-]|%3D|%2F|%2B)+)?|["
+                "\\-]{5}BEGIN(?:[a-z\\s]|%20)+PRIVATE(?:\\s|%20)KEY[\\-]{5}[^\\-]+[\\-]{5}END(?:[a-z\\s]|%"
+                "20)+PRIVATE(?:\\s|%20)KEY|ssh-rsa(?:\\s|%20)*(?:[a-z0-9\\/\\.+]|%2F|%5C|%2B){100,}",
+            },
+            {"name": "DD_TRACE_OTEL_ENABLED", "origin": "unknown", "value": False},
+            {"name": "DD_TRACE_PROPAGATION_STYLE_EXTRACT", "origin": "unknown", "value": "tracecontext ,datadog"},
+            {"name": "DD_TRACE_PROPAGATION_STYLE_INJECT", "origin": "unknown", "value": "tracecontext ,datadog"},
```
Comment (mabdinur) [inline tests/telemetry/test_writer.py:104]:
```suggestion
            {"name": "DD_TRACE_PROPAGATION_STYLE_EXTRACT", "origin": "unknown", "value": "tracecontext,datadog"},
            {"name": "DD_TRACE_PROPAGATION_STYLE_INJECT", "origin": "unknown", "value": "tracecontext,datadog"},
```

File: tests/telemetry/test_writer.py [hunk 17]
```diff
@@ -218,7 +257,7 @@ def test_app_client_configuration_changed_event(telemetry_writer, test_agent_ses
     """asserts that queuing a configuration sends a valid telemetry request"""
 
     telemetry_writer.add_configuration("appsec_enabled", True)
-    telemetry_writer.add_configuration(TELEMETRY_PROPAGATION_STYLE_EXTRACT, "datadog")
+    telemetry_writer.add_configuration("DD_TRACE_PROPAGATION_STYLE_EXTRACT", "datadog")
```
Comment (ZStriker19) [inline tests/telemetry/test_writer.py:260]: ah thanks for this correction

**Review by ZStriker19 — APPROVED** [review]

**Pull request merged**